common/sphinx.c

```diff
@@ -563,9 +590,17 @@ struct route_step *process_onionpacket(
 		return tal_free(step);
 
 	deserialize_hop_data(&step->hop_data, paddedheader);
+	sphinx_parse_payload(step, paddedheader);
 
-	/* Extract how many frames we need to shift away */
-	shift_size = ((paddedheader[0] >> 4) + 1) * FRAME_SIZE;
+	/* Extract how many bytes we need to shift away */
+	if (paddedheader[0] == 0x00) {
+		shift_size = FRAME_SIZE;
+	} else {
+		/* In addition to the raw payload we need to also shift the
+		 * length encoding itself and the HMAC away. */
+		vsize = varint_get(paddedheader, 3, &shift_size);
+		shift_size += vsize + HMAC_SIZE;
```

Comment:
This is wrong on several levels.

1. I worry what happens if we don't have 3 bytes here, can we be tricked into reading off the end?
2. This uses bitcoin's varint and we are now using BigSize which is big-endian.
3. shift_size needs to be sanity checked carefully!

Reply:
> 1. I worry what happens if we don't have 3 bytes here, can we be tricked into reading off the end?
No, we have `paddedheader[2*ROUTING_INFO_SIZE];`, i.e., 2600 bytes here.

> 2. This uses bitcoin's varint and we are now using BigSize which is big-endian.
Yep, fixing up

> 3. shift_size needs to be sanity checked carefully!
I went and just abort when `shift_size >= ROUTING_INFO_SIZE` since that'd shift out all non-filler infos and leave us with garbage at the end anyway :-)